Rework allow_reuse in decorators

adriangb opened the pull request:
Closes #4436.

This implements pretty much the same behavior that pyright and ruff/flake8 (via `F811`) enforce.

dmontagu commented on tests/test_serialize.py:
tests look good

dmontagu commented on pydantic/main.py:
```diff
@@ -157,6 +166,10 @@ def hash_func(self_: Any) -> int:
 
+    @classmethod
+    def __prepare__(cls, *args: Any, **kwargs: Any) -> Mapping[str, object]:
+        return _ModelNamespaceDict()
```
this is very neat

dmontagu commented on pydantic/main.py:
```diff
@@ -56,6 +56,15 @@
+    def __setitem__(self, k: str, v: object) -> None:
+        existing: Any = self.get(k, None)
+        if existing and v is not existing and isinstance(existing, _decorators.PydanticDecoratorMarker):
+            warnings.warn(f'`{k}` overrides an existing Pydantic `{existing.decorator_info.decorator_repr}` decorator')
```
Do we care about the case of `elif existing and v is not existing and isinstance(v, _decorators.PydanticDecoratorMarker)`? I guess seems okay to not cover it.

adriangb replied:
I don't think so. I don't want us to become a runtime linter lol, so I'm only covering the case of overriding a name with one of our decorators

dmontagu replied:
lol sounds good

dmontagu replied:
I guess the only thing I'll mention is that I think the original purpose of this was kind of runtime linting lol. So it might be a good idea to get Samuel's eyes on this and confirm he's in agreement. But I think we can merge with this decision and just change later.

adriangb replied:
I mean it still does runtime linting, just in a limited fashion. cc @samuelcolvin

dmontagu commented on pydantic/decorators.py:
```diff
@@ -298,6 +295,8 @@ def root_validator(
     before or after standard model parsing/validation is performed.
     """
+    if allow_reuse is True:  # pragma: no cover
+        warn('`allow_reuse` is deprecated and will be ignored', DeprecationWarning)
```
```suggestion
        warn('`allow_reuse` is deprecated and will be ignored; it should no longer be necessary', DeprecationWarning)
```
Can phrase this differently, but I think a good idea to tell people that their code will probably "just work" if they drop the `allow_reuse`, clarifying that it's not deprecated and will break their code.

dmontagu replied:
(And same note above)

dmontagu reviewed (commented):
Looks like a big improvement. Couple notes but nothing major, I'd say merge when you feel ready

adriangb commented:
@dmontagu feedback addressed

samuelcolvin commented on pydantic/main.py:
```diff
@@ -56,6 +56,15 @@
 _base_class_defined = False
 
 
+class _ModelNamespaceDict(dict):  # type: ignore[type-arg]
```
please can we add a docstring explaining what this does?

samuelcolvin reviewed (commented):
looks great, thank you.

My only concern is whether we should keep the parameter and raise an error or warning, rather than remove it?

adriangb commented:
> My only concern is whether we should keep the parameter and raise an error or warning, rather than remove it?

I kept it for the V1 decorators and added a warning. But I removed it for the V2 decorators, I think a minor breaking change like that in an alpha is okay.
